src/ripple/basics/Number.h

```diff
@@ -0,0 +1,342 @@
+    explicit Number(rep mantissa, int exponent);
+    explicit constexpr Number(rep mantissa, int exponent, unchecked) noexcept;
+
+    Number(XRPAmount const& x);
+
+    constexpr rep
+    mantissa() const noexcept;
+    constexpr int
+    exponent() const noexcept;
+
+    constexpr Number
+    operator+() const noexcept;
+    constexpr Number
+    operator-() const noexcept;
+    Number&
+    operator++();
+    Number
+    operator++(int);
+    Number&
+    operator--();
+    Number
+    operator--(int);
+
+    Number&
+    operator+=(Number const& x);
+    Number&
+    operator-=(Number const& x);
+
+    Number&
+    operator*=(Number const& x);
+    Number&
+    operator/=(Number const& x);
+
+    explicit operator XRPAmount() const;  // round to nearest, even on tie
+    explicit operator rep() const;        // round to nearest, even on tie
+
+    friend constexpr bool
+    operator==(Number const& x, Number const& y) noexcept
+    {
+        return x.mantissa_ == y.mantissa_ && x.exponent_ == y.exponent_;
+    }
+
+    friend constexpr bool
+    operator!=(Number const& x, Number const& y) noexcept
+    {
+        return !(x == y);
+    }
+
+    friend constexpr bool
+    operator<(Number const& x, Number const& y) noexcept
+    {
+        // If the two amounts have different signs (zero is treated as positive)
+        // then the comparison is true iff the left is negative.
+        bool const lneg = x.mantissa_ < 0;
+        bool const rneg = y.mantissa_ < 0;
+
+        if (lneg != rneg)
+            return lneg;
+
+        // Both have same sign and the left is zero: the right must be
+        // greater than 0.
+        if (x.mantissa_ == 0)
+            return y.mantissa_ > 0;
+
+        // Both have same sign, the right is zero and the left is non-zero.
+        if (y.mantissa_ == 0)
+            return false;
+
+        // Both have the same sign, compare by exponents:
+        if (x.exponent_ > y.exponent_)
+            return lneg;
+        if (x.exponent_ < y.exponent_)
+            return !lneg;
+
+        // If equal exponents, compare mantissas
+        return x.mantissa_ < y.mantissa_;
+    }
+
+    friend constexpr bool
+    operator>(Number const& x, Number const& y) noexcept
+    {
+        return y < x;
+    }
+
+    friend constexpr bool
+    operator<=(Number const& x, Number const& y) noexcept
+    {
+        return !(y < x);
+    }
+
+    friend constexpr bool
+    operator>=(Number const& x, Number const& y) noexcept
+    {
+        return !(x < y);
+    }
+
+    friend std::ostream&
+    operator<<(std::ostream& os, Number const& x)
+    {
+        return os << to_string(x);
+    }
+
+    // Thread local rounding control.  Default is to_nearest
+    enum rounding_mode { to_nearest, towards_zero, downward, upward };
+    static rounding_mode
+    getround();
+    // Returns previously set mode
+    static rounding_mode
+    setround(rounding_mode mode);
+
+private:
+    static thread_local rounding_mode mode_;
```

Comment:
I'm open to suggestions here.

It looks like co-routines resuming on different threads is an explicit design for the co-routine author, and not something that happens by accident.  So we could just document "don't do that" for the `Number` class.

Changing `rounding_mode` to be `static` doesn't seem like a good idea because that means all computations using `Number` (on different threads) will be forced to share the same rounding mode.

Changing `rounding_mode` to be a non-static data member isn't practical as there will be many `Number` instances participating in a single expression.

Reply:
It setting the rounding mode is used exactly once in non-test code: in AMM_formula.h. How would the code look if we made the rounding mode a template parameter and made it part of the type instead of a thread local? It sounds like all but one place uses the same rounding mode. If this is true, that might work out well.

Reply:
Handy utility for setting rounding mode, and resetting it back to previous:

https://github.com/HowardHinnant/rippled/blob/NumberMerge/src/test/basics/Number_test.cpp#L29-L45

Used like this:

https://github.com/HowardHinnant/rippled/blob/NumberMerge/src/test/basics/Number_test.cpp#L359

Reply:
If `to_nearest` is the only mode ever set, we don't need this facility because `to_nearest` is the default, and what was in place prior to the introduction of `Number::setround`.